### ISSUE TITLE
Migrate away from Github

### ISSUE THREAD

**Issue by martanne** [martanne/vis#495]
For some reason Github decided to flag my account.

* Git hosting: [repo.or.cz/vis.git](http://repo.or.cz/vis.git)
* Project communication: mailing list hosted by [FreeLists](https://www.freelists.org/) and IRC channel
* Bug tracker: ?
* Continuous integration with:
   - Travis (for Linux and macOS bulids, CoverityScan and Codecov integration)
   - AppVeyor (for Windows builds)

  We could probably use a read-only github project as mirror to keep the build hooks working.

**Comment by george-b** [comment]
For what it's worth I reported this to Github support yesterday but was told the owner of the account (yourself) would have to report it. Oddly durring this time I could `git pull` from Github.

**Comment by martanne** [comment]
Apparently my account was mistaken for a spam bot. I tried to contact them twice through the [web form](https://github.com/contact) while being logged in with the flagged account, but they seemingly never got my messages. Finally, I sent a mail to support@github.com, once they got my inquiry they then promptly cleared the flag.

**Comment by george-b** [comment]
> Also I think they're behavior (redirecting all requests to 404 pages) is quite bad. At the very least they should have displayed a message that the account is temporarily suspended due to violations of the terms of service or similar.

Agreed it was quite confusing.

**Comment by aeosynth** [comment]
despite this incident, i would vote to stay on github. if that's not possible, there are other similar services which provide integrated hosting/bug tracker, eg [bitbucket], [gitlab], [gogs].

[bitbucket]: https://bitbucket.org/
[gitlab]: https://gitlab.com
[gogs]: https://gogs.io/

**Comment by lanodan** [comment]
I’d "vote" to keep github only as a mirror/hub, we should learn from mistakes and try to not give too much power to a unknow organisation.

**Comment by josuah** [comment]
At least, having http://repo.or.cz/vis.git up to date reveals to be useful for these kind of situations.

**Comment by halfwit** [comment]
I agree with the idea of keeping a mirror, as redundancy is always a welcome feature. http://repo.or.cz/vis.git and lists are certainly less clunky than the github UI for people who use the web in different manners as well. (Text browsers, etc)

**Comment by apotheon** [comment]
Keeping a mirror on GitHub is a good idea -- better than keeping the main repository on GitHub, actually (especially if you can still take pull requests from GitHub).  Among all the obvious reasons to keep a GitHub mirror, it also provides a certain amount of advertising for other approaches to hosting Git repositories online, thus enhancing awareness of ways people can re-decentralize the DVCSes that people have been centralizing on GitHub for years.

**Comment by aeosynth** [comment]
is this still happening?

**Comment by mqudsi** [comment]
Moving away from GitHub unfortunately means losing a lot of potential new contributors. It's become somewhat of an (unfortunate) de facto.

**Comment by lanodan** [comment]
@mqudsi:
I think we are more like thinking about using github as a hub (mirror the repos) and try to move other things (issues, wiki, patches, …).

Maybe I’m a bit too activist and I completely disagree by “it’s that way and let’s keep it” that I’m understanding by using the de facto state of github (which might go down/close/… any time).

**Comment by mcepl** [comment]
I think this bug should be either closed or followed upon.

**Comment by 3dc1d3** [comment]
In the spirit of suckless tools (which, to me, vis is part of) I'll mention here [fossil-scm](https://fossil-scm.org) in case it's strikes others as a potential move away from GitHub:

> Git provides file versioning services only, whereas Fossil adds an integrated wiki, ticketing & bug tracking, embedded documentation, technical notes, and a web forum, all within a single nicely-designed skinnable web UI, protected by a fine-grained role-based access control system. These additional capabilities are available for Git as 3rd-party add-ons, but with Fossil they are integrated into the design. One way to describe Fossil is that it is **"GitHub-in-a-box."**

It's a single executable that can be hosted on anything, written by the same great minds behind SQLite.  

It has a `git-export` feature which can serve for GitHub mirroring etc. 

[Fossil vs Git](https://fossil-scm.org/home/doc/trunk/www/fossil-v-git.wiki)
[Fossil coming from Git](https://fossil-scm.org/fossil/doc/trunk/www/gitusers.md)

This is not a push, just an idea I think is worth exploring.

**Comment by apotheon** [comment]
I host my own Fossil repositories and use [fossgit](https://rubygems.org/gems/fossgit) to mirror them to GitHub.  I like the idea, though I could easily imagine there being logistical issues with taking that approach.  At the moment, fossgit is only a one-way mirroring tool.  Maybe it'll acquire the ability to sync in the other direction this year.

**Comment by mcepl** [comment]
Fossil is not a suckless solution, it is the suckfull one. In this Blessed Year of Our Lord 2020 it makes absolutely no sense to use any other versioning system than the one which name consists of letters 'g', 'i', and 't' and no other. Which wouldn’t be a problem for non-Github, there are plenty of self-hosting git solutions, but the issues are not repository itself, but issues and pull requests.

For those there are basically two solutions. First, going full suckless way (to quote dwm webpage: “Because dwm is customized through editing its source code, it’s pointless to make binary packages of it. This keeps its userbase small and elitist. No novices asking stupid questions.”) and use the combination of IRC and email list for support and git-request-pull(1) for pull requests.

If you want “normal” issue tracker and something like pull requests, than the only viable alternative is gitlab. Yes, it is not perfect, but it is an established alternative, it is mostly free source, and it is possible to export repository (including issue tracker and merge requests). Anything else (and mostly anything else than gitlab.com) is just combination of all disadvantages of gitlab with all isolation of non-mainstream solution.

My suggestion would be either to move to gitlab.com, or just hold your nose and continue using GitHub. However, it is better for the hygiene of the issue tracker, if some decision is made and this ticket is closed one or other way.

**Comment by apotheon** [comment]
You kinda have an attitude problem, @mcept.  You don't even make a good case for your position.  I don't know how the vis maintainer feels about it, but I at least think you might try relying more on reason than whatever emotional reaction you're having as a means of making a case for something.

I think GitLab is an acceptable idea, too, though not for any reasons you really made clear -- because you basically never bothered to make any good reasons clear in your comment.  If I was a knee-jerk reaction kind of person, you'd have convinced me that GitLab was a *bad* idea, just because of how you presented your opinions.

**Comment by RibalGZ** [comment]
There is already a sourcehut mirror: https://git.sr.ht/~martanne/vis
In addition, it was mentioned that it would be interesting to integrate sourcehut with github (see [759](https://github.com/martanne/vis/issues/759#issuecomment-590268055))

**Comment by 3dc1d3** [comment]
thank you @apotheon.  btw fossil's native git integration is pretty decent now and works both ways.  there is the occasional new corner case but of course the forums are very responsive and helpful.

**Comment by apotheon** [comment]
@3dc1d3: Yeah, fossgit just makes it easier for some use cases, in part by providing a straightforward way to use configuration to simplify the actual process, which is why I want it for my own purposes.

@RibalGZ: Strange.  I had no idea sourcehut existed.  It seems to offer a bit less utility than GitLab, apart from the fact there's evidently a Mercurial option as well (which is particularly interesting in an age when even Bitbucket dropped Mercurial support -- despite offering some advantages over Git).

**Comment by martanne** [comment]
By now there also exists a [vis project on sourcehut](https://sr.ht/~martanne/vis/) with an [associated mailing list](https://lists.sr.ht/~martanne/devel).

I know some community members prefer an email based workflow, so feel free to send patches that way.

**Comment by mcepl** [comment]
OK, sourcehut looks tentatively interesting. I would have to play with it in anger to see how it really works, but somebody finally using git-send-mail(1) cannot be that bad.

**Comment by apotheon** [comment]
Some of the weird plans for payments and so on with sourcehut seem a bit . . . well, weird.

**Comment by mcepl** [comment]
> Some of the weird plans for payments and so on with sourcehut seem a bit . . . well, weird.

Well, I prefer when they are honest about their business plan, and they actually have some business plan. And given that it is [`git send-email`](https://git-send-email.io/)-based system, contributors actually do not have to have repositories on the server.
